### PR TITLE
[ENH] Add cross-validation scoring for image-based decoders

### LINKED ISSUE #6068: [ENH] create a `cross_val_score` compatible with decoders, frem etc. to ease decoding accuracy computation.
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Describe your proposed enhancement in detail.

implement a `cross_val_decoder_score` analogous to cross_val_score, that would take images as input.

### Benefits to the change

Much simpler access to performance metrics for decoders.

### Pseudocode for the new behavior, if applicable

```python
# insert your code below

```

### PR DESCRIPTION
**Pull request by mohammedsadeghi**
- Closes #6068

Use of AI tools:
- [x] LLM tools were used to generate at least part of the content of this pull-request.
- [ ] No LLM tools were used to generate any part of the content of this pull-request.

Changes proposed in this pull request:

- Add `nilearn.decoding.cross_val_decoder_score` to accept volume and surface images directly. sklearn's `cross_val_score` interprets a 4D image's first spatial dimension as its sample count and rejects a `SurfaceImage`; converting them into per-sample image lists allows the helper to delegate cross-validation to sklearn.
- Fit a fresh decoder within each outer training fold and score held-out images. The decoder's internal CV remains inside that training fold, so the returned scores measure outer held-out performance separately from `cv_scores_`.
- Validate volume collections before cross-validation and reject items containing multiple volumes. Preserve surface input shapes during conversion and document sklearn scoring, grouped inner/outer CV, metadata routing, and parallelism.
- Showcase held-out scoring in the Haxby ANOVA/SVM example, keeping runs separate in both CV loops. Add regression tests, the public API listing, the updated API discovery counts, and a changelog entry.

Validation:

- 52 helper tests and 11 API discovery tests plus four doctests passed with the current minimum dependencies, including scikit-learn 1.6.0. Coverage includes all four decoder classes, valid and invalid image representations, manual fold equivalence, group routing, training-only masker fits, input preservation, error handling, and parallel execution.
- The complete Haxby ANOVA/SVM example ran successfully on the cached dataset and returned finite scores for all three outer folds.
- The strict no-plot documentation build passed the modified example; the local build stopped later because this machine has no LaTeX executable for unrelated math-image rendering.
- All applicable pre-commit checks passed.

### CODE REVIEW

File: nilearn/decoding/_validation.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    >>> rng = np.random.default_rng(0)
+    >>> img = Nifti1Image(rng.normal(size=(5, 5, 5, 30)), np.eye(4))
+    >>> mask = Nifti1Image(np.ones((5, 5, 5)), np.eye(4))
+    >>> decoder = Decoder(mask=mask, cv=2, screening_percentile=None)
+    >>> scores = cross_val_decoder_score(
+    ...     decoder, img, np.tile([0, 1], 15), cv=3, scoring="accuracy"
+    ... )
+    >>> scores.shape
+    (3,)
+    """
```
Comment (bthirion) [inline nilearn/decoding/_validation.py:143]: We probably need more checks. i.e. users should not supply list of 4D images.

**Review by bthirion — COMMENTED** [review]
Really cool addition !
I believe it should be showcased in an example of the decoding section.

File: nilearn/decoding/_validation.py (hunk 1 above)
Comment (mohammedsadeghi) [reply nilearn/decoding/_validation.py:143]: Done.

**Comment by Remi-Gau** [comment]
converting to draft so we do not merge for the 0.14.1 release (this is a new feature so cannot go in before 0.15.0)

**Comment by Remi-Gau** [comment]
We should probably illustrate that in an example (maybe by simply update one of the existing example), no?

**Comment by mohammedsadeghi** [comment]
@Remi-Gau 
Tell me what to do and I'll deliver it ASAP!

**Comment by bthirion** [comment]
> @Remi-Gau Tell me what to do and I'll deliver it ASAP!

`plot_haxby_different_estimators` or
`plot_haxby_anova_svm` could benefit from this.

**Review by bthirion — APPROVED** [review]
LGTM, thx.

File: examples/02_decoding/plot_haxby_anova_svm.py [hunk 2]
```diff
@@ -104,12 +104,42 @@
...
+scores = cross_val_decoder_score(
+    decoder,
+    func_img,
+    conditions,
+    cv=GroupKFold(n_splits=3),
+    groups=run_label,
+    params={"groups": run_label},
+    scoring="accuracy",
+)
+print(f"Held-out run accuracies: {scores}")
```
Comment (Remi-Gau) [inline examples/02_decoding/plot_haxby_anova_svm.py:139]:
```suggestion

# %%
print(f"Held-out run accuracies: {scores}")
```

**Review by Remi-Gau — APPROVED** [review]

File: nilearn/decoding/_validation.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    Scoring, validation, metadata routing, and parallel execution follow
+    scikit-learn's ``cross_val_score`` conventions. Existing sample lists
+    can also be passed directly to that function.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from nibabel import Nifti1Image
+    >>> from nilearn.decoding import Decoder, cross_val_decoder_score
+    >>> rng = np.random.default_rng(0)
```
Comment (Remi-Gau) [inline nilearn/decoding/_validation.py:134]:
```suggestion
    >>> import numpy as np
    >>> from nibabel import Nifti1Image
    >>>
    >>> from nilearn.decoding import Decoder, cross_val_decoder_score
    >>>
    >>> rng = np.random.default_rng(0)
```

File: nilearn/decoding/_validation.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        accepted. Splitters receive a list of image samples as ``X``.
+
+    n_jobs : :obj:`int` or None, default=None
+        Number of outer folds to evaluate in parallel. None uses one job
+        unless a joblib parallel configuration specifies otherwise.
+        -1 uses all processors. Set the decoder's own ``n_jobs=1`` when
+        parallelizing outer folds to avoid nested parallelism.
+
+    verbose : :obj:`int`, default=0
+        Verbosity of scikit-learn's cross-validation computation.
```
Comment (Remi-Gau) [inline nilearn/decoding/_validation.py:80]:
```suggestion
    %(verbose0)s
```

File: nilearn/decoding/_validation.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        of folds; None uses five folds. Classification with binary or
+        multiclass targets uses stratified folds, otherwise K-fold is used.
+        A splitter or an iterable of (train, test) index pairs is also
+        accepted. Splitters receive a list of image samples as ``X``.
+
+    n_jobs : :obj:`int` or None, default=None
+        Number of outer folds to evaluate in parallel. None uses one job
+        unless a joblib parallel configuration specifies otherwise.
+        -1 uses all processors. Set the decoder's own ``n_jobs=1`` when
+        parallelizing outer folds to avoid nested parallelism.
```
Comment (Remi-Gau) [inline nilearn/decoding/_validation.py:77]:
```suggestion
    %(n_jobs)s
```

File: nilearn/decoding/_validation.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+def cross_val_decoder_score(
+    estimator,
+    imgs,
+    y,
+    *,
+    groups=None,
+    scoring=None,
+    cv=None,
+    n_jobs=None,
```
Comment (Remi-Gau) [inline nilearn/decoding/_validation.py:19]:
```suggestion
    n_jobs: NJobs=1,
```

NJobs must be imported from nilearn.nilearn_typing

File: nilearn/decoding/_validation.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
+"""Cross-validation with image inputs."""
+
+import numpy as np
+from sklearn.model_selection import cross_val_score
+
+from nilearn.image import check_niimg_3d, iter_img
+from nilearn.nilearn_typing import NiimgLike
+from nilearn.surface import SurfaceImage
+
+
```
Comment (Remi-Gau) [inline nilearn/decoding/_validation.py:10]:
```suggestion
@fill_doc
```

decorator needs to be imported from nilearn._utils.docs

File: nilearn/decoding/_validation.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+        When scikit-learn metadata routing is enabled, pass groups only in
+        ``params`` and configure the estimator's metadata requests.
+
+    scoring : :obj:`str`, callable, or None, default=None
+        A scikit-learn scorer name or a callable with signature
+        ``scorer(estimator, imgs, y)`` returning a single score.
+        If None, use the estimator's ``score`` method. For decoders this uses
+        their own ``scoring`` parameter.
+
+    cv : :obj:`int`, cross-validation splitter, iterable, or None, default=None
```
Comment (Remi-Gau) [inline nilearn/decoding/_validation.py:66]: there is a common cv attributes definition in nilearn._utils.docs can you check if it can be reused here?

File: nilearn/decoding/_validation.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+def cross_val_decoder_score(
+    estimator,
+    imgs,
+    y,
+    *,
+    groups=None,
+    scoring=None,
+    cv=None,
+    n_jobs=None,
+    verbose=0,
```
Comment (Remi-Gau) [inline nilearn/decoding/_validation.py:20]:
```suggestion
    verbose: Verbose=0,
```
Verbose must be imported from nilearn.nilearn_typing

File: nilearn/decoding/_validation.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+    >>> rng = np.random.default_rng(0)
+    >>> img = Nifti1Image(rng.normal(size=(5, 5, 5, 30)), np.eye(4))
+    >>> mask = Nifti1Image(np.ones((5, 5, 5)), np.eye(4))
+    >>> decoder = Decoder(mask=mask, cv=2, screening_percentile=None)
+    >>> scores = cross_val_decoder_score(
+    ...     decoder, img, np.tile([0, 1], 15), cv=3, scoring="accuracy"
+    ... )
+    >>> scores.shape
+    (3,)
+    """
```
Comment (Remi-Gau) [inline nilearn/decoding/_validation.py:143]:
```suggestion
    """
    check_params(locals())
```
must import check_params from nilearn/_utils/param_validation.py

this allows to validate several input parameters
